File: engine/src/main/java/com/terralite/engine/simulation/SimulationTick.java
package com.terralite.engine.simulation;

import java.time.Duration;
import java.util.Objects;

public record SimulationTick(long index, Duration delta, Duration totalTime) {
    public SimulationTick {
        if (index <= 0) {
            throw new IllegalArgumentException("Tick index must be positive");
        }
        Objects.requireNonNull(delta, "delta");
        Objects.requireNonNull(totalTime, "totalTime");
        if (delta.isZero() || delta.isNegative()) {
            throw new IllegalArgumentException("Tick delta must be positive");
        }
        if (totalTime.isNegative()) {
            throw new IllegalArgumentException("Total time cannot be negative");
        }
    }
}

File: engine/src/commonMain/kotlin/com/terralite/engine/simulation/SimulationTick.kt
package com.terralite.engine.simulation

/**
 * Represents a single simulation step.
 *
 * @param index     1-based tick counter.
 * @param delta     Duration of this tick in nanoseconds. Must be positive.
 * @param totalTime Accumulated simulation time in nanoseconds since start. Must be non-negative.
 */
@JvmRecord
data class SimulationTick(val index: Long, val delta: Long, val totalTime: Long) {
    init {
        require(index > 0) { "Tick index must be positive" }
        require(delta > 0) { "Tick delta must be positive" }
        require(totalTime >= 0) { "Total time cannot be negative" }
    }
}
